### ISSUE TITLE
Gamescope: cursor escapes nested window

### ISSUE THREAD

**Issue by VeilSilence** [hyprwm/Hyprland#5076]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit a065b481f32f4297596a9cd9c538794fd4669b1a dirty (cursormgr: use XCURSOR_THEME for x themes).
Date: Mon Mar 11 23:33:26 2024
Tag: v0.36.0-100-ga065b481

flags: (if any)
```

</details>

### Bug or Regression?

Bug

### Description

As title says, cursor easily escapes from gamescope window at any games I tested, basically. Manually setting "fullscreen" or start gamescope with `-f` not helping much.
In sway and plasma 6, cursor stays inside gamescope. 
My mouse related settings:
<details><summary>Details</summary>
<p>

follow_mouse = 1
env = WLR_NO_HARDWARE_CURSORS,1

</p>
</details> 

### How to reproduce

1: Run for example glxgears `gamescope --force-grab-cursor glxgears`.
2: Mouse easily escape the nested session.

**Comment by bfalk8** [comment]
Performing a `git bisect` on the code lead me to find that [this commit](https://github.com/hyprwm/Hyprland/commit/d72ea5f2a7fdb0d0a7bf914412327195b05199b0) is what caused the change in behavior. @VeilSilence can you confirm this on your end?

**Comment by sawb** [comment]
I also have this issue on the latest versions.

Downgrading to a cached package at version v0.33.1-37-g4190b967 for now has resolved the issue.

**Comment by vaxerski** [comment]
dam

**Comment by vaxerski** [comment]
check the above, why does gamescope have to be cringe

**Comment by sawb** [comment]
I just tried with the above commit, and it is easier to escape the window now, making the issue worse. The cursor will move to the left display when it is slightly to the left of the center of gamescope. 

Additionally, some applications such as Steam,Vesktop,Feishin are no longer drawing on the display despite running. Not sure if that was caused by this commit.

**Comment by vaxerski** [comment]
> Additionally, some applications such as Steam,Vesktop,Feishin are no longer drawing on the display despite running. Not sure if that was caused by this commit.

it was and I just fixed it

> I just tried with the above commit, and it is easier to escape the window now, making the issue worse. The cursor will move to the left display when it is slightly to the left of the center of gamescope.

cap? I am unable to escape it at all with the patch.

**Comment by sawb** [comment]
This is what currently happens for me on the latest commit : https://streamable.com/roo50v
This is an old version (v0.33.1-37-g4190b967) where the mouse is locking: https://streamable.com/cnbhht

I'm using the following arguments in gamescope:
gamescope -W 3440 -H 1440 -w 3440 -h 1440 -f -r 330 -o 15 --force-grab-cursor

**Comment by vaxerski** [comment]
offtopic but why are you launching cs2 in gamescope?

**Comment by sawb** [comment]
Couple reasons:

Makes it easy to manipulate the resolution (stretch/black bars etc)
I had a few issues running CS2 without gamescope (mouse bugs, resolution not right etc). Didn't want to try the cs2 related plugin since gamescope is good enough for me.

I don't personally feel a difference between the synced gamescope and tearing non-gamescope cs2 to bother running it without. Although I definitely will enable tearing if that ever comes to gamescope.

**Comment by vaxerski** [comment]
gamescope doesn't need tearing support, just make its framerate unlimited and make hl tear it.

Anyways, still can't repro, tho I'm testing with dolphin and not cs

**Comment by sawb** [comment]
Also happens to me on every other game/program I've tried including glxgears and vkcube. 

I'm wondering if maybe there is a difference between our DPI/sensitivity settings that is allowing me to escape and not you. I'm using 1600dpi and input:sensitivity= -0.4.

**Comment by bfalk8** [comment]
To help illustrate the issue, here is the basic example of gamescope with the '--force-grab-cursor' running the glxgears app with the latest codebase (commit 86dc46ffea2fa72abe1228c1fad1b42d1507e45d of Hyprland) where the cursor doesn't get constrained:
https://streamable.com/zm6fjc

And the old behavior (captured from v0.35.0 of Hyprland) where the cursor is successfully constrained:
https://streamable.com/7z90t7

I have also produced a more minimal Hyprland config that still causes this issue to occur on the latest codebase if it helps
```
monitor=,3840x2160@120,auto,auto

misc {
    disable_hyprland_logo = true
    disable_splash_rendering = true
}

input {
    follow_mouse = 1
}

# gamescope windows float
windowrulev2 = float, class:^(gamescope)$
windowrulev2 = noborder, class:^(gamescope)$
windowrulev2 = noshadow, class:^(gamescope)$

$mainMod = ALT

# BINDINGS #
bind = $mainMod, Q, killactive, 
bind = $mainMod, M, exit, 
bind = $mainMod, V, togglefloating, 
bind = $mainMod, Return, exec, alacritty
bind = $mainMod, r, exec, gamescope --force-grab-cursor glxgears
```
My exact use case that lead me here is that I will wrap RTS games (i.e. AoE2: DE, Starcraft, Stormgate, etc...) in gamescope as a floating window on my 48 inch monitor and use gamescope's cursor grab functionality to keep my cursor contained so i can still "pan the camera" in game.

The output of `hyprctl systeminfo` if it is helpful (the system is currently using the RX6800 XT gpu):
```
Hyprland, built from branch HEAD at commit 84ab8d11e8951a6551d1e1bf87796a8589da6d47  (props: bump ver to 0.35.0).
Date: Mon Feb 5 01:59:02 2024
Tag: v0.35.0

flags: (if any)


System Information:
System name: Linux
Node name: brandon-tower
Release: 6.8.1-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Sat, 16 Mar 2024 17:15:35 +0000


GPU information:
0b:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Lexa PRO [Radeon 540/540X/550/550X / RX 540X/550/550X] [1002:699f] (rev c7) (prog-if 00 [VGA controller])
0e:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 21 [Radeon RX 6800/6800 XT / 6900 XT] [1002:73bf] (rev c1) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:
```

**Comment by vaxerski** [comment]
> running the glxgears app with the latest codebase (commit https://github.com/hyprwm/Hyprland/commit/86dc46ffea2fa72abe1228c1fad1b42d1507e45d of Hyprland) where the cursor doesn't get constrained:

from your systeminfo:
> Tag: v0.35.0

I am afraid you're on a _very_ outdated commit.

**Comment by bfalk8** [comment]
Sorry, `v0.35.0`  is the version I rolled back to after realizing the new version wasn't constraining my cursor. I redid my videos with newer versions of the codebase.

Video from commit 328ab43 (the last commit before the new behavior was introduced):

https://github.com/hyprwm/Hyprland/assets/11812541/89d895cb-b26f-4bb1-bb9d-812ca6214a17

Associated systeminfo:
```
Hyprland, built from branch HEAD at commit 328ab431655f207a5ee59737291f8e8d85fab992 dirty (hyprpm: don't copy .so if file doesn't exist).
Date: Sat Mar 2 11:06:09 2024
Tag: v0.36.0-31-g328ab431

flags: (if any)


System Information:
System name: Linux
Node name: brandon-tower
Release: 6.8.1-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Sat, 16 Mar 2024 17:15:35 +0000


GPU information: 
0b:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Lexa PRO [Radeon 540/540X/550/550X / RX 540X/550/550X] [1002:699f] (rev c7) (prog-if 00 [VGA controller])
0e:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 21 [Radeon RX 6800/6800 XT / 6900 XT] [1002:73bf] (rev c1) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

```


Video from commit 6b28bf5 (a recent commit that shows the new behavior):

https://github.com/hyprwm/Hyprland/assets/11812541/915859a2-4e1e-4cc0-b104-1472edb09c27

Associated systeminfo:
```
Hyprland, built from branch main at commit 6b28bf563e3b4e5f6631985f24c88e8a8ebaa0d4 dirty (keybinds: Fix exit trigger by moving it to monitor.frame (5240)).
Date: Sun Mar 24 18:46:59 2024
Tag: v0.37.1-63-g6b28bf56

flags: (if any)


System Information:
System name: Linux
Node name: brandon-tower
Release: 6.8.1-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Sat, 16 Mar 2024 17:15:35 +0000


GPU information: 
0b:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Lexa PRO [Radeon 540/540X/550/550X / RX 540X/550/550X] [1002:699f] (rev c7) (prog-if 00 [VGA controller])
0e:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 21 [Radeon RX 6800/6800 XT / 6900 XT] [1002:73bf] (rev c1) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

```

new `hyprlandd.conf` debug config used for both videos
```
monitor=,3840x2160@120,auto,auto

misc {
    disable_hyprland_logo = true
    disable_splash_rendering = true
}

input {
    follow_mouse = 1
}

# gamescope windows float
windowrulev2 = float, class:^(gamescope)$
# windowrulev2 = noborder, class:^(gamescope)$
# windowrulev2 = noshadow, class:^(gamescope)$

$mainMod = ALT

# BINDINGS #
bind = $mainMod, Q, killactive, 
bind = $mainMod, M, exit, 
bind = $mainMod, V, togglefloating, 
bind = $mainMod, Return, exec, alacritty
bind = $mainMod, r, exec, gamescope --force-grab-cursor glxgears

# WORKSPACE BINDINGS #
# Move focus with mainMod + vim keys
bind = $mainMod, h, movefocus, l
bind = $mainMod, l, movefocus, r
bind = $mainMod, k, movefocus, u
bind = $mainMod, j, movefocus, d

# Swap windows with mainMod + SHIFT + vim keys
bind = $mainMod SHIFT, h, swapwindow, l
bind = $mainMod SHIFT, l, swapwindow, r
bind = $mainMod SHIFT, k, swapwindow, u
bind = $mainMod SHIFT, j, swapwindow, d

# Switch workspaces with mainMod + [0-9]
bind = $mainMod, 1, workspace, 1
bind = $mainMod, 2, workspace, 2
bind = $mainMod, 3, workspace, 3

# Move active window to a workspace with mainMod + SHIFT + [0-9]
bind = $mainMod SHIFT, 1, movetoworkspace, 1
bind = $mainMod SHIFT, 2, movetoworkspace, 2
bind = $mainMod SHIFT, 3, movetoworkspace, 3

# Move/resize windows with mainMod + LMB/RMB and dragging
bindm = $mainMod, mouse:272, movewindow
bindm = $mainMod, mouse:273, resizewindow
```

Hope this helps!

**Comment by sawb** [comment]
> @vaxerski I didn't bother bisecting the commits, but some change went into hyprland or gamescope over the past few days that fixed this issue. Running both libraries on their latest commits has produced the desired behavior.
> 
> Thanks for taking the time to look at this! I really appreciate the work you have done/are doing on the hyprland project, it really is the window manager I have wanted for a long time <3
> 
> @VeilSilence @sawb may want to confirm, but I would be fine with closing this issue.

I have found the issue is resolved by using gamescope-git instead of the arch-provided package. So on the latest hyprland commit (https://github.com/hyprwm/Hyprland/commit/93d05114716e847c37f49d3cc2d0c5cb01d06a24) , the cursor can escape with the arch gamescope package, but not the built gamescope-git from latest source. Seems like gamescope has done something to resolve this on their end.

**Comment by VeilSilence** [comment]
Sorry for late reponse. NixOS take a bit of time to figure for me.
I tested on hyprland commit d8429eebc66e7091381502edc0fc7335790f62e4 and git gamescope commit. Now cursor working same as in KDE/sway so far.

Found are small thing in comparison between KDE and hyprland. 
`gamemoderun gamescope -w 2560 -h 1440  -f -r 240 --mangoapp --force-grab-cursor  %command%` 
Game: Warhammer 40,00: Soulstorm. Mangohud is not visible on Hyprland. In KDE, I see Mangohud overlay. 
`gamemoderun gamescope -w 2560 -h 1440  -f -r 240 mangohud --force-grab-cursor  %command%` i can see MangoHud in both compositors. Only in this game for reason...

Other than that, we can close the issue i think.
